opennms-config-dao/thresholding/impl/src/main/java/org/opennms/netmgt/config/dao/thresholding/impl/OnmsThresholdingDao.java

```diff
@@ -66,14 +87,66 @@ public void saveConfig() {
         saveableConfigContainer.saveConfig();
     }
 
+    /**
+     * @return the merged configuration consisting of the filesystem configuration and any configuration provided by
+     * extensions
+     */
     @Override
-    public ThresholdingConfig getConfig() {
+    public ThresholdingConfig getReadOnlyConfig() {
+        return getMergedConfig();
+    }
+
+    /**
+     * @return just the configuration from the filesystem since configuration provided by extensions is read only
+     */
+    @Override
+    public ThresholdingConfig getWriteableConfig() {
         return saveableConfigContainer.getConfig();
     }
 
+    private synchronized ThresholdingConfig getMergedConfig() {
+        ThresholdingConfig externalConfig = extContainer.getObject();
+
+        if (filesystemConfig == null && externalConfig == null) {
+            return null;
+        } else if (externalConfig == null) {
+            return filesystemConfig;
+        } else if (filesystemConfig == null) {
+            return externalConfig;
+        }
+
+        // Create a merged config by combining the config from filesystem and the external config provided by extensions
+        ThresholdingConfig mergedConfig = new ThresholdingConfig();
+
+        List<Group> groups = new ArrayList<>();
+        groups.addAll(filesystemConfig.getGroups());
+        groups.addAll(externalConfig.getGroups());
+        mergedConfig.setGroups(Collections.unmodifiableList(groups));
+
+        return mergedConfig;
+    }
+
+    private synchronized void publishMergedConfig() {
+        try {
+            jsonStore.put(JSON_STORE_KEY, objectMapper.writeValueAsString(getMergedConfig()),
+                    ConfigDaoConstants.JSON_KEY_STORE_CONTEXT);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
     @Override
     public void reload() {
         saveableConfigContainer.reload();
-        super.reload();
+    }
+
+    @Override
+    public void onConfigChanged() {
+        publishMergedConfig();
+    }
+
+    private synchronized void fileSystemConfigUpdated(ThresholdingConfig updatedConfig) {
+        filesystemConfig = updatedConfig;
```

Comment:
I'm not super happy about using this `filesystemConfig` reference but it was the quickest/easiest way to fix the test failures.

The issue is that in the constructor of this class, we create a `FileSystemSaveableConfigContainer ` and pass it a callback that eventually triggers `publishMergedConfig()` which would previously do a `saveableConfigContainer.getConfig()` to retrieve the config. However this would fail since that object had not finished initializing (since the callback is called as part of init of that object). So now it uses this new reference which is always kept up to date via the callback.

The main reason I don't like it is that there is now two ways to of tracking the file system state in these objects, via the `saveableConfigContainer.getConfig()` and via the `filesystemConfig` reference. Care must be taken to use the `filesystemConfig` one when merging config for the first time and `saveableConfigContainer.getConfig()` when serving the config for modifications to make sure users mutate the right object.

If you see an easy way to fix this let me know, otherwise it does seem to be working fine.